### PR TITLE
Track and report peer response size/time

### PR DESCRIPTION
**Pull request by gsalgado**
Closes: #1128

### CODE REVIEW

File: trinity/protocol/common/managers.py [hunk 1]
```diff
@@ -33,6 +35,26 @@
...
+        total_msgs = len(self.responses)
+        total_response_time = sum([item[0] for item in self.responses])
+        total_items = sum([item[1] for item in self.responses])
+        avg_rtt = total_response_time / total_msgs
+        per_item_rtt = total_response_time / total_items
+        return 'count=%d, avg_rtt=%.2f, avg_time_per_item=%.5f' % (
+            total_msgs, avg_rtt, per_item_rtt)
+
+    def add(self, time: float, size: int) -> None:
+        self.responses.append((time, size))
```
Comment (pipermerriam) [inline trinity/protocol/common/managers.py]: This looks like it has `O(n)` storage/memory footprint.  Unless I'm missing something we need to aggregate these stats using a running average and total count of number of responses.

File: p2p/peer.py [hunk 2]
```diff
@@ -1056,21 +1063,17 @@ def _test() -> None:
             peer_pool.logger.info("Waiting for peer connection...")
             await asyncio.sleep(0.2)
         peer = peer_pool.highest_td_peer
-        block_hash = decode_hex(
-            '0x59af08ab31822c992bb3dad92ddb68d820aa4c69e9560f07081fa53f1009b152')
+        headers = await peer.requests.get_block_headers(2440319, max_headers=100)  # type: ignore
+        hashes = [header.hash for header in headers]
```
Comment (pipermerriam) [inline p2p/peer.py]: clever!

File: trinity/protocol/common/managers.py (hunk 1 above)
Comment (gsalgado) [reply trinity/protocol/common/managers.py]: I thought this wouldn't be a problem as it's just a two-tuple for every *response* (not every msg received) and I've rarely seen peer connections lasting more than a few hours. Do you think we need to deal with this now or would a separate, low-priority issue be enough?

Comment (pipermerriam) [reply trinity/protocol/common/managers.py]: > Do you think we need to deal with this now or would a separate, low-priority issue be enough?

That works for me.  Can you add it to the next milestone to be sure we address it quickly.

Comment (gsalgado) [reply trinity/protocol/common/managers.py]: I'll actually do it here and save an unnecessary PR roundtrip

Comment (gsalgado) [reply trinity/protocol/common/managers.py]: Done

**Comment by pipermerriam** [comment]
Just thought of adding a timeout counter to this as well to track number of failed requests.

File: trinity/protocol/common/managers.py [hunk 3]
```diff
@@ -33,6 +34,27 @@
...
+    def __init__(self) -> None:
+        self.total_msgs = 0
+        self.total_items = 0
+        self.total_response_time = 0.0
+
+    def get_stats(self) -> str:
+        if not self.total_msgs:
+            return 'None'
+        avg_rtt = self.total_response_time / self.total_msgs
+        per_item_rtt = self.total_response_time / self.total_items
```
Comment (pipermerriam) [inline trinity/protocol/common/managers.py]: This is succeptable to division by zero if the peer is returning empty responses.

Comment (gsalgado) [reply trinity/protocol/common/managers.py]: I've changed the stats to include the total number of items as well, and use 0 for per_item_rtt when self.total_items is 0

**Pull request merged**